### PR TITLE
fix: make apollo-template compile in typescript

### PR DESCRIPTION
**Pull request by ankitjena**
Fix for #122
- add tsconfig.json, build script and configuration
- Fix typescript compile errors
- Create Context interface for resolver context
```
export interface Context {
  pubsub: RedisPubSub
  req: express.Request
  db: knex<any, unknown[]>
}
```
- Typings in resolvers
```
key: (_: any, args: any, context: Context) => { }
```

### CODE REVIEW

**Comment by wtrocki** [comment]
So this is the place where I would love to see cli being used to generate code and at least check if it will compile

File: templates/apollo-starter-ts/tslint.json [hunk 1]
```diff
@@ -14,7 +14,8 @@
       "object-literal-sort-keys": false,
       "trailing-comma": false,
       "object-literal-shorthand": false,
-      "no-console": false
+      "no-console": false,
+      "interface-name": false
```
Comment (wtrocki) [inline templates/apollo-starter-ts/tslint.json]: Reason?

File: templates/apollo-starter-ts/package.json [hunk 2]
```diff
@@ -3,9 +3,11 @@
   "version": "0.3.0",
   "description": "",
   "private": true,
-  "main": "dist/index.js",
+  "main": "dist/src/index.js",
   "scripts": {
-    "start": "ts-node src/index.ts"
+    "start": "node dist/src/index.js",
+    "build": "tsc",
+    "dev": "ts-node src/index.ts"
```
Comment (wtrocki) [inline templates/apollo-starter-ts/package.json]: I will add the watch option.

File: packages/graphback/src/resolvers/knex/resolverImplementation.ts [hunk 3]
```diff
@@ -2,7 +2,7 @@ const knexContext = `context.db`
 const pubsub = `context.pubsub`
 
 export const createTemplate = (fieldName: string, tableName: string, typeName: string): string => {
-    return `${fieldName}: async (_, args, context) => {
+    return `${fieldName}: async (_: any, args: any, context: Context) => {
```
Comment (wtrocki) [inline packages/graphback/src/resolvers/knex/resolverImplementation.ts]: Where is context going to be defined? Use something like `GraphBackContext | any`

**Comment by wtrocki** [comment]
We need additional script - can be npm or bash that will execute command line tool on build and generate project. Then we should check if the generated project compiles (follow up will be if it actually works)

File: packages/graphback/src/resolvers/knex/resolverImplementation.ts (hunk 3 above)
Comment (ankitjena) [reply packages/graphback/src/resolvers/knex/resolverImplementation.ts]: It will be defined in the context.ts in the template

**Comment by ankitjena** [comment]
> We need additional script - can be npm or bash that will execute command line tool on build and generate project. Then we should check if the generated project compiles (follow up will be if it actually works)

When we run npm run build at the root of the project, it also executes the build commands in templates. If there is any error, the build will fail. Thus we can see if the template actually compiles. As we have made changes for the type error in resolvers

File: packages/graphback/src/resolvers/resolverTemplate.ts [hunk 4]
```diff
@@ -1,7 +1,11 @@
 import { TargetResolverContext } from './knex/targetResolverContext';
 
+const imports = `import { Context } from '../src/context'`
```
Comment (ankitjena) [inline packages/graphback/src/resolvers/resolverTemplate.ts]: Doing the import here

File: templates/apollo-starter-ts/tslint.json (hunk 1 above)
Comment (ankitjena) [reply templates/apollo-starter-ts/tslint.json]: Demanding resolver name to start with "I"

File: packages/graphback/src/resolvers/resolverTemplate.ts (hunk 4 above)
Comment (wtrocki) [reply packages/graphback/src/resolvers/resolverTemplate.ts]: Rename it to something like GeneratedContext or something similar.

**Comment by wtrocki** [comment]
> When we run npm run build at the root of the project, it also executes the build commands in templates. If there is any error, the build will fail. Thus we can see if the template actually compiles. As we have made changes for the type error in resolvers

Yes, but do we generated resolvers on it? This is very simple task we just need to generate resolvers on top of the sample template to see if this compiles as whole (t did not before so let's do it)

**Comment by ankitjena** [comment]
> > When we run npm run build at the root of the project, it also executes the build commands in templates. If there is any error, the build will fail. Thus we can see if the template actually compiles. As we have made changes for the type error in resolvers
> 
> Yes, but do we generated resolvers on it? This is very simple task we just need to generate resolvers on top of the sample template to see if this compiles as whole (t did not before so let's do it)

Then we would need a example model file to generate resolvers. Not sure if I am getting this right but we can have script like
```
graphback db && tcs
```
So we generate resolvers and then compile. Just the template will need to have a model by default.

**Comment by wtrocki** [comment]
I think we can just copy model from CLI right?

**Comment by ankitjena** [comment]
> I think we can just copy model from CLI right?
> 
> WDYT:
> 
> ```
> cp {model} ./model
> graphback generate
> graphback db 
> ```

I think we should do this in integration test in cli. And if we do strict typing on generated resolvers it will compile as well. Plus it will make the build command the user can reuse, which won't alter the generated files.

**Pull request merged**